pkg/util/rankedset/rankedset.go

```diff
@@ -0,0 +1,182 @@
+}
+
+func (*NoRank) Rank() int64 {
+	return 0
+}
+
+func New() *RankedSet {
+	return &RankedSet{
+		ranked: btree.New(32),
+		set:    make(map[string]*treeItem),
+	}
+}
+
+// Add inserts the item into the set.
+// If an item with the same Key existed in the set, it is removed and returned.
+func (s *RankedSet) Add(item SetItem) SetItem {
+	old := s.Remove(item)
+
+	key := item.Key()
+	value := &treeItem{item: item}
+
+	s.ranked.ReplaceOrInsert(value) // should always return nil because we call remove first
+	s.set[key] = value
+
+	return old
+}
+
+// Remove deletes the item from the set based on Key (Rank is ignored).
+// The removed item is returned if it existed in the set.
+func (s *RankedSet) Remove(item SetItem) SetItem {
+	key := item.Key()
+	value, ok := s.set[key]
+	if !ok {
+		return nil
+	}
+
+	s.ranked.Delete(value) // should always return the same data as value (non-nil)
+	delete(s.set, key)
+
+	return value.item
+}
+
+func (s *RankedSet) Min() SetItem {
+	if min := s.ranked.Min(); min != nil {
+		return min.(*treeItem).item
+	}
+	return nil
+}
+
+func (s *RankedSet) Max() SetItem {
+	if max := s.ranked.Max(); max != nil {
+		return max.(*treeItem).item
+	}
+	return nil
+}
+
+func (s *RankedSet) Len() int {
+	return len(s.set)
+}
+
+func (s *RankedSet) Get(item SetItem) SetItem {
+	if value, ok := s.set[item.Key()]; ok {
+		return value.item
+	}
+	return nil
+}
+
+func (s *RankedSet) Has(item SetItem) bool {
+	_, ok := s.set[item.Key()]
+	return ok
+}
+
+// List returns all items in the set in ranked order.
+// If remove is set to true, the returned items are removed from the set.
+func (s *RankedSet) List(remove bool) []SetItem {
+	return s.ascend(
+		func(item SetItem) bool {
+			return true
+		},
+		remove,
+	)
+}
+
+// LessThan returns all items less than the given rank in ranked order.
+// If remove is set to true, the returned items are removed from the set.
+func (s *RankedSet) LessThan(rank int64, remove bool) []SetItem {
+	return s.ascend(
+		func(item SetItem) bool {
+			return item.Rank() < rank
+		},
+		remove,
+	)
+}
+
+// setItemIterator allows callers of ascend to iterate in-order over the set.
+// When this function returns false, iteration will stop.
+type setItemIterator func(item SetItem) bool
+
+func (s *RankedSet) ascend(iterator setItemIterator, remove bool) []SetItem {
+	var items []SetItem
+	s.ranked.Ascend(func(i btree.Item) bool {
+		item := i.(*treeItem).item
+		if !iterator(item) {
+			return false
+		}
+		items = append(items, item)
+		return true
+	})
+	// remove after Ascend since it is probably not safe to delete while iterating
+	if remove {
+		for _, item := range items {
+			s.Remove(item)
+		}
+	}
+	return items
+}
+
+var _ btree.Item = &treeItem{}
+
+type treeItem struct {
+	item SetItem
+}
+
+func (i *treeItem) Less(than btree.Item) bool {
+	other := than.(*treeItem).item
+
+	selfKey := i.item.Key()
+	otherKey := other.Key()
+
+	// !a.Less(b) && !b.Less(a) means a == b
+	if selfKey == otherKey {
```

Comment:
This is not right.
if you test for equality first it means you will arbitrarily replace a matching key element with the new one on insertion, regardles of the value of it's rank member.
This will result in insertion in the wrong place so the set will not be ranked according to rank.
You must assume the higher level functions properly dealt with preventing duplicates in the btree, and not care for key equality at all.

Reply:
Note you do not see this condition happening in the test because currently the higher level functions always remove a mathing key before adding a new one, so this case is effectively never happening, currently.